components/prism-bash.js

```diff
@@ -19,66 +29,168 @@
 					punctuation: /\(\(?|\)\)?|,|;/
 				}
 			},
-			// Command Substitution
+			// [1]: Command Substitution
 			{
-				pattern: /\$\([^)]+\)|`[^`]+`/,
+				pattern: /\$\((?:\([^)]+\)|[^()])+\)|`[^`]+`/,
 				greedy: true,
 				inside: {
 					variable: /^\$\(|^`|\)$|`$/
 				}
 			},
-			/\$(?:[\w#?*!@]+|\{[^}]+\})/i
-		]
+			// [2]: Brace expansion
+			{
+				pattern: /\$\{[^}]+\}/,
+				greedy: true,
+				inside: {
+					operator: /:[-=?+]?|[!\/]|##?|%%?|\^\^?|,,?/,
+					punctuation: /[\[\]]/,
+					environment: {
+						pattern: new RegExp("(\\{)(?:" + envVarList + ")(?=$|\\}|\\W)"),
+						lookbehind: true,
+						alias: 'constant'
+					}
+				}
+			},
+			/\$(?:\w+|[#?*!@$])/
+		],
+		// man echo / printf, + \"
+		entity: /\\([abceEfnrtv\\"]|O?[0-7]{1,3}|x[0-9a-fA-F]{1,2}|u[0-9a-fA-F]{4}|U[0-9a-fA-F]{8})/
 	};
 
 	Prism.languages.bash = {
-		'shebang': {
-			pattern: /^#!\s*\/bin\/bash|^#!\s*\/bin\/sh/,
+		shebang: {
+			pattern: /^#!\s*\/.*/,
 			alias: 'important'
 		},
-		'comment': {
-			pattern: /(^|[^"{\\])#.*/,
+		comment: {
+			pattern: /(^|[^"{\\$])#.*/,
 			lookbehind: true
 		},
-		'string': [
-			//Support for Here-Documents https://en.wikipedia.org/wiki/Here_document
+		'function-name': [
+			// a) function foo {
+			// b) foo() {
+			// c) function foo() {
+			// but not “foo {”
 			{
-				pattern: /((?:^|[^<])<<\s*)["']?(\w+?)["']?\s*\r?\n(?:[\s\S])*?\r?\n\2/,
+				// a) and c)
+				pattern: /((^|\W)function\s+)\w+(?=(\s*\(\s*\))?\s*\{)/,
+				lookbehind: true,
+				alias: 'function'
+			},
+			{
+				// b)
+				pattern: /(^|\W)\w+(?=\s*\(\s*\)\s*\{)/,
+				lookbehind: true,
+				alias: 'function'
+			}
+		],
+		// Highlight variable names as variables in for and select beginnings.
+		'for-or-select': {
+			pattern: /((?:^|\W)(?:for|select)\s+)\w+(?=\s+in\s)/,
+			alias: 'variable',
+			lookbehind: true
+		},
+		// Highlight variable names as variables in the left-hand part
+		// of assignments (“=” and “+=”).
+		'assign-left': {
+			pattern: /(^|[\s;|&]|[<>]\()\w+(?=\+?=)/,
+			inside: {
+				environment: {
+					pattern: new RegExp("(^|[\\s;|&]|[<>]\\()(?:" + envVarList + ")(?=$|\\W)"),
+					lookbehind: true,
+					alias: 'constant'
+				}
+			},
+			alias: 'variable',
+			lookbehind: true
+		},
+		string: [
+			// Support for Here-documents https://en.wikipedia.org/wiki/Here_document
+			{
+				pattern: /((?:^|[^<])<<-?\s*)(\w+?)\s*(?:\r?\n|\r)(?:[\s\S])*?(?:\r?\n|\r)\2/,
 				lookbehind: true,
 				greedy: true,
 				inside: insideString
 			},
+			// Here-document with quotes around the tag
+			// → No expansion (so no “inside”).
+			{
+				pattern: /((?:^|[^<])<<-?\s*)(["'])(\w+?)\2\s*(?:\r?\n|\r)(?:[\s\S])*?(?:\r?\n|\r)\3/,
```

Comment:
Why is `\w+?` reluctant?
It's always followed by either `"` or `'`, so why not let it be greedy?

Reply:
I guess you're right. Probably copypasted that from the “without quotes” version, and my lack of confidence prevented me from fiddling with the regex too much.